### PR TITLE
[expo-notifications] Move handling implementation to NotificationsService delegate

### PR DESCRIPTION
**Pull request by sjchmiela**
# Why

A prerequisite to single-threaded notifications handling.

# How

More refactoring, moving staff from `NotificationsHelper` to appropriate `NotificationsService`' delegates. This time to `ExpoHandlingDelegate`.

In the meantime I have also removed a memory leak introduced in https://github.com/expo/expo/pull/10640 — we don't need to _observe_ the state, we can check for the state every time we need to access it.

# Test Plan

Push and scheduled notifications are presented.

### CODE REVIEW

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/ExpoHandlingDelegate.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+     * your listener, use [NotificationManager] singleton module.
+     *
+     *
+     * Purposefully the argument is expected to be a [NotificationManager] and just a listener.
+     *
+     *
+     * This class doesn't hold strong references to listeners, so you need to own your listeners.
+     *
+     * @param listener A listener instance to be informed of new push device tokens.
+     */
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/ExpoHandlingDelegate.kt:35]:
```suggestion
    /**
     * Used only by [NotificationManager] instances. If you look for a place to register
     * your listener, use [NotificationManager] singleton module.
     *
     * Purposefully the argument is expected to be a [NotificationManager] and just a listener.
     *
     * This class doesn't hold strong references to listeners, so you need to own your listeners.
     *
     * @param listener A listener instance to be informed of new push device tokens.
     */
```

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/ExpoHandlingDelegate.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        sListenersReferences[listener] = listenerReference
+        if (!sPendingNotificationResponses.isEmpty()) {
+          val responseIterator = sPendingNotificationResponses.iterator()
+          while (responseIterator.hasNext()) {
+            listener.onNotificationResponseReceived(responseIterator.next())
+            responseIterator.remove()
+          }
+        }
+      }
+    }
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/ExpoHandlingDelegate.kt:49]:
```suggestion
      if (sListenersReferences.containsKey(listener)) {
      	// this listener has already been registered
        return
      }

      sListenersReferences[listener] = WeakReference(listener)

      if (!sPendingNotificationResponses.isEmpty()) {
        val responseIterator = sPendingNotificationResponses.iterator()
        while (responseIterator.hasNext()) {
          listener.onNotificationResponseReceived(responseIterator.next())
          responseIterator.remove()
        }
      }
    }
```

**Review by bbarthec — APPROVED** [review]
What would you say about being more verbose in naming the interfaces? 🤔 
- `HandlingDelegate` ➡️ `NotificationHandlingDelegate`,
- `PresentationDelegate` ➡️ `NotificationPresentationDelegate`?

I mean, we don't have any strong verbosity convention, so it's just a suggestion for thinking it over 😉 

Nevertheless, PR looks good 😉 I really like the gradual migration from `java` to `kotlin` 👍

**Comment by sjchmiela** [comment]
Good call on the interface names, I'll change all the interfaces names in a separate pull request. 👍

**Pull request merged**